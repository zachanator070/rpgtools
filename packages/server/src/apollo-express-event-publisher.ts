import { EventPublisher } from "./types";
import { PubSub } from "apollo-server-express";
import { injectable } from "inversify";

@injectable()
export class ApolloExpressEventPublisher implements EventPublisher {
	pubsub = new PubSub();

	asyncIterator(events: string[]): AsyncIterator<any> {
		return this.pubsub.asyncIterator(events);
	}

	publish(event: string, payload: any): Promise<void> {
		return this.pubsub.publish(event, payload);
	}
}
