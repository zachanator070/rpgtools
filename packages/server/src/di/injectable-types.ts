import {DatabaseContext} from "../dal/database-context";

export const INJECTABLE_TYPES = {

	DomainEntity: Symbol.for("DomainEntity"),

	// domain entities
	Article: Symbol.for("Article"),
	Chunk: Symbol.for("Chunk"),
	File: Symbol.for("File"),
	Game: Symbol.for("Game"),
	Image: Symbol.for("Image"),
	Item: Symbol.for("Item"),
	Model: Symbol.for("Model"),
	Monster: Symbol.for("Monster"),
	PermissionAssignment: Symbol.for("PermissionAssignment"),
	Person: Symbol.for("Person"),
	Pin: Symbol.for("Pin"),
	Place: Symbol.for("Place"),
	Role: Symbol.for("Role"),
	ServerConfig: Symbol.for("ServerConfig"),
	User: Symbol.for("User"),
	WikiFolder: Symbol.for("WikiFolder"),
	World: Symbol.for("World"),
	Event: Symbol.for("Event"),
	Calendar: Symbol.for("Calendar"),

	// entity factories
	ArticleFactory: Symbol.for("ArticleFactory"),
	AclFactory: Symbol.for('AclFactory'),
	ChunkFactory: Symbol.for("ChunkFactory"),
	FileFactory: Symbol.for("FileFactory"),
	GameFactory: Symbol.for("GameFactory"),
	ImageFactory: Symbol.for("ImageFactory"),
	ItemFactory: Symbol.for("ItemFactory"),
	ModelFactory: Symbol.for("ModelFactory"),
	MonsterFactory: Symbol.for("MonsterFactory"),
	PersonFactory: Symbol.for("PersonFactory"),
	PinFactory: Symbol.for("PinFactory"),
	PlaceFactory: Symbol.for("PlaceFactory"),
	RoleFactory: Symbol.for("RoleFactory"),
	ServerConfigFactory: Symbol.for("ServerConfigFactory"),
	UserFactory: Symbol.for("UserFactory"),
	WikiFolderFactory: Symbol.for("WikiFolderFactory"),
	WikiPageFactory: Symbol.for("WikiPageFactory"),
	WorldFactory: Symbol.for("WorldFactory"),
	EventWikiFactory: Symbol.for("EventFactory"),
	CalendarFactory: Symbol.for("CalendarFactory"),

	CharacterFactory: Symbol.for('CharacterFactory'),
	CharacterAttributeFactory: Symbol.for('CharacterAttributeFactory'),
	PathNodeFactory: Symbol.for('NodePathFactory'),
	StrokeFactory: Symbol.for('StrokeFactory'),
	FogStrokeFactory: Symbol.for('FogStrokeFactory'),
	MessageFactory: Symbol.for('MessageFactory'),
	InGameModelFactory: Symbol.for('InGameModelFactory'),

	AgeFactory: Symbol.for("AgeFactory"),
	MonthFactory: Symbol.for("MonthFactory"),
	DayOfTheWeekFactory: Symbol.for("DayOfTheWeekFactory"),

	// repositories
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	FileRepository: Symbol.for("FileRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	EventWikiRepository: Symbol.for("EventWikiRepository"),
	CalendarRepository: Symbol.for("CalendarRepository"),
	StrokeRepository: Symbol.for('StrokeRepository'),
	FogStrokeRepository: Symbol.for('FogStrokeRepository'),

	SqlPermissionControlledRepository: Symbol.for('SqlPermissionControlledRepository'),

	// authorization policies
	ArticleAuthorizationPolicy: Symbol.for("ArticleAuthorizationPolicy"),
	ChunkAuthorizationPolicy: Symbol.for("ChunkAuthorizationPolicy"),
	FileAuthorizationPolicy: Symbol.for("FileAuthorizationPolicy"),
	GameAuthorizationPolicy: Symbol.for("GameAuthorizationPolicy"),
	ImageAuthorizationPolicy: Symbol.for("ImageAuthorizationPolicy"),
	ItemAuthorizationPolicy: Symbol.for("ItemAuthorizationPolicy"),
	ModelAuthorizationPolicy: Symbol.for("ModelAuthorizationPolicy"),
	MonsterAuthorizationPolicy: Symbol.for("MonsterAuthorizationPolicy"),
	PersonAuthorizationPolicy: Symbol.for("PersonAuthorizationPolicy"),
	PinAuthorizationPolicy: Symbol.for("PinAuthorizationPolicy"),
	PlaceAuthorizationPolicy: Symbol.for("PlaceAuthorizationPolicy"),
	RoleAuthorizationPolicy: Symbol.for("RoleAuthorizationPolicy"),
	ServerConfigAuthorizationPolicy: Symbol.for("ServerConfigAuthorizationPolicy"),
	UserAuthorizationPolicy: Symbol.for("UserAuthorizationPolicy"),
	WikiFolderAuthorizationPolicy: Symbol.for("WikiFolderAuthorizationPolicy"),
	WikiPageAuthorizationPolicy: Symbol.for("WikiPageAuthorizationPolicy"),
	WorldAuthorizationPolicy: Symbol.for("WorldAuthorizationPolicy"),
	CalendarAuthorizationPolicy: Symbol.for("CalendarAuthorizationPolicy"),
	FogStrokeAuthorizationPolicy: Symbol.for('FogStrokeAuthorizationPolicy'),
	StrokeAuthorizationPolicy: Symbol.for('StrokeAuthorizationPolicy'),

	// archive repositories
	ArchiveArticleRepository: Symbol.for("ArchiveArticleRepository"),
	ArchiveChunkRepository: Symbol.for("ArchiveChunkRepository"),
	ArchiveGameRepository: Symbol.for("ArchiveGameRepository"),
	ArchiveImageRepository: Symbol.for("ArchiveImageRepository"),
	ArchiveItemRepository: Symbol.for("ArchiveItemRepository"),
	ArchiveModelRepository: Symbol.for("ArchiveModelRepository"),
	ArchiveMonsterRepository: Symbol.for("ArchiveMonsterRepository"),
	ArchivePermissionAssignmentRepository: Symbol.for("ArchivePermissionAssignmentRepository"),
	ArchivePersonRepository: Symbol.for("ArchivePersonRepository"),
	ArchivePinRepository: Symbol.for("ArchivePinRepository"),
	ArchivePlaceRepository: Symbol.for("ArchivePlaceRepository"),
	ArchiveRoleRepository: Symbol.for("ArchiveRoleRepository"),
	ArchiveServerConfigRepository: Symbol.for("ArchiveServerConfigRepository"),
	ArchiveUserRepository: Symbol.for("ArchiveUserRepository"),
	ArchiveWikiFolderRepository: Symbol.for("ArchiveWikiFolderRepository"),
	ArchiveWikiPageRepository: Symbol.for("ArchiveWikiPageRepository"),
	ArchiveWorldRepository: Symbol.for("ArchiveWorldRepository"),
	ArchiveFileRepository: Symbol.for("ArchiveFileRepository"),

	// services
	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	RoleService: Symbol.for("RoleService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	// session contexts
	SessionContextFactory: Symbol.for("SessionContextFactory"),
	SecurityContextFactory: Symbol.for("SecurityContextFactory"),

	// 5e
	Open5eApiClient: Symbol.for("Open5eApiClient"),
	Dnd5eApiClient: Symbol.for("Dnd5eApiClient"),

	// delta
	DeltaFactory: Symbol.for("DeltaFactory"),

	// cache
	Cache: Symbol.for("Cache"),

	// archive
	ArchiveFactory: Symbol.for("ArchiveFactory"),

	ZipArchiveFactory: Symbol.for("ZipArchiveFactory"),
	ZipArchive: Symbol.for('ZipArchive'),

	// event publisher
	EventPublisher: Symbol.for("EventPublisher"),

	// server
	ApiServer: Symbol.for("ApiServer"),
	ServerProperties: Symbol.for("ServerProperties"),
	RpgToolsServer: Symbol.for('RpgToolsServer'),
	DbEngine: Symbol.for("DbEngine"),
	SqlDbEngine: Symbol.for('SqlDbEngine'),

	// data loaders
	ArticleDataLoader: Symbol.for("ArticleDataLoader"),
	ChunkDataLoader: Symbol.for("ChunkDataLoader"),
	FileDataLoader: Symbol.for("FileDataLoader"),
	GameDataLoader: Symbol.for("GameDataLoader"),
	ImageDataLoader: Symbol.for("ImageDataLoader"),
	ItemDataLoader: Symbol.for("ItemDataLoader"),
	ModelDataLoader: Symbol.for("ModelDataLoader"),
	MonsterDataLoader: Symbol.for("MonsterDataLoader"),
	PermissionAssignmentDataLoader: Symbol.for("PermissionAssignmentDataLoader"),
	PersonDataLoader: Symbol.for("PersonDataLoader"),
	PinDataLoader: Symbol.for("PinDataLoader"),
	PlaceDataLoader: Symbol.for("PlaceDataLoader"),
	RoleDataLoader: Symbol.for("RoleDataLoader"),
	ServerConfigDataLoader: Symbol.for("ServerConfigDataLoader"),
	UserDataLoader: Symbol.for("UserDataLoader"),
	WikiFolderDataLoader: Symbol.for("WikiFolderDataLoader"),
	WikiPageDataLoader: Symbol.for("WikiPageDataLoader"),
	WorldDataLoader: Symbol.for("WorldDataLoader"),
	CalendarDataLoader: Symbol.for('CalendarDataLoader'),
	EventDataLoader: Symbol.for('EventDataLoader'),

	// seeders
	RoleSeeder: Symbol.for("RoleSeeder"),
	ServerConfigSeeder: Symbol.for("ServerConfigSeeder"),

	// database context
	DatabaseContext: Symbol.for("DatabaseContext"),
	DatabaseContextFactory: Symbol.for("DatabaseContextFactory"),

	// repo mapper
	EntityMapper: Symbol.for("EntityMapper"),

	// mongodb
	FilterFactory: Symbol.for("FilterFactory"),
	MongoDbMigrationV40: Symbol.for("MongoDbMigrationV40"),


};
