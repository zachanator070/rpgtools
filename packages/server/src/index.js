import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
