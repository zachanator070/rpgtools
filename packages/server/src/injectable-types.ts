export const INJECTABLE_TYPES = {
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PermissionAssignmentRepository: Symbol.for("PermissionAssignmentRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	FileRepository: Symbol.for("FileRepository"),

	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	SessionContextFactory: Symbol.for("SessionContextFactory"),
	Open5eApiClient: Symbol.for("Open5eApiClient"),

	Cache: Symbol.for("Cache"),
};
