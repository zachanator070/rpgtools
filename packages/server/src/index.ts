import {container} from "./di/inversify";
import {INJECTABLE_TYPES} from "./di/injectable-types";
import {ApiServer} from "./types";

const server = container.get<ApiServer>(INJECTABLE_TYPES.ApiServer);

(async () => {
	await server.start();
})();
