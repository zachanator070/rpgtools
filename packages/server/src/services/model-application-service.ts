import {
	AuthorizationService,
	Cache,
	EventPublisher,
	Factory,
	FileFactory,
	ModelFactory,
	ModelRepository,
	ModelService,
	PermissionAssignmentFactory,
} from "../types";
import { MODEL_ADMIN, MODEL_RW } from "../../../common/src/permission-constants";
import { MODEL } from "../../../common/src/type-constants";
import { inject, injectable } from "inversify";
import { INJECTABLE_TYPES } from "../di/injectable-types";
import { ModelAuthorizationRuleset } from "../security/ruleset/model-authorization-ruleset";
import { SecurityContext } from "../security/security-context";
import { FileUpload } from "graphql-upload";
import { Model } from "../domain-entities/model";
import { FilterCondition } from "../dal/filter-condition";
import { DbUnitOfWork } from "../dal/db-unit-of-work";
import { GAME_MODEL_DELETED } from "../resolvers/subscription-resolvers";

@injectable()
export class ModelApplicationService implements ModelService {
	@inject(INJECTABLE_TYPES.Cache)
	cache: Cache;

	@inject(INJECTABLE_TYPES.EventPublisher)
	eventPublisher: EventPublisher;

	@inject(INJECTABLE_TYPES.AuthorizationService)
	authorizationService: AuthorizationService;

	@inject(INJECTABLE_TYPES.ModelRepository)
	modelRepository: ModelRepository;

	@inject(INJECTABLE_TYPES.ModelAuthorizationRuleset)
	modelAuthorizationRuleset: ModelAuthorizationRuleset;

	@inject(INJECTABLE_TYPES.DbUnitOfWorkFactory)
	dbUnitOfWorkFactory: Factory<DbUnitOfWork>;

	@inject(INJECTABLE_TYPES.FileFactory)
	fileFactory: FileFactory;
	@inject(INJECTABLE_TYPES.ModelFactory)
	modelFactory: ModelFactory;
	@inject(INJECTABLE_TYPES.PermissionAssignmentFactory)
	permissionAssignmentFactory: PermissionAssignmentFactory;

	createModel = async (
		context: SecurityContext,
		worldId: string,
		name: string,
		fileUpload: FileUpload,
		depth: number,
		width: number,
		height: number,
		notes: string
	) => {
		const unitOfWork = this.dbUnitOfWorkFactory();
		const world = await unitOfWork.worldRepository.findById(worldId);
		if (!world) {
			throw new Error(`World with id ${worldId} does not exist`);
		}

		if (!(await this.modelAuthorizationRuleset.canCreate(context, world))) {
			throw new Error("You do not have permission to add models to this world");
		}

		fileUpload = await fileUpload;

		const file = this.fileFactory(null, fileUpload.filename, fileUpload.createReadStream(), null);
		await unitOfWork.fileRepository.create(file);
		const model = this.modelFactory(
			null,
			worldId,
			name,
			depth,
			width,
			height,
			file.filename,
			file._id,
			notes
		);
		await unitOfWork.modelRepository.create(model);
		for (let permission of [MODEL_RW, MODEL_ADMIN]) {
			const permissionAssignment = this.permissionAssignmentFactory(
				null,
				permission,
				model._id,
				MODEL
			);
			await unitOfWork.permissionAssignmentRepository.create(permissionAssignment);
			context.user.permissions.push(permissionAssignment._id);
			context.permissions.push(permissionAssignment);
		}
		await unitOfWork.userRepository.update(context.user);
		await unitOfWork.commit();
		return model;
	};

	updateModel = async (
		context: SecurityContext,
		modelId: string,
		name: string,
		depth: number,
		width: number,
		height: number,
		notes: string,
		file?: FileUpload
	) => {
		const unitOfWork = this.dbUnitOfWorkFactory();
		const model = await unitOfWork.modelRepository.findById(modelId);
		if (!model) {
			throw new Error(`Model with id ${modelId} does not exist`);
		}
		if (!(await this.modelAuthorizationRuleset.canWrite(context, model))) {
			throw new Error("You do not have permission to edit this model");
		}
		if (file) {
			if (await this.cache.exists(model.fileName)) {
				await this.cache.delete(model.fileName);
			}

			if (model.fileId) {
				const file = await unitOfWork.fileRepository.findById(model.fileId);
				await unitOfWork.fileRepository.delete(file);
			}
			file = await file;

			const newFile = this.fileFactory(null, file.filename, file.createReadStream(), null);
			await unitOfWork.fileRepository.create(newFile);
			model.fileId = newFile._id;
			model.fileName = file.filename;
		}
		model.name = name;
		model.depth = depth;
		model.width = width;
		model.height = height;
		model.notes = notes;
		await unitOfWork.modelRepository.update(model);
		await unitOfWork.commit();
		return model;
	};

	deleteModel = async (context: SecurityContext, modelId: string) => {
		const unitOfWork = this.dbUnitOfWorkFactory();
		const model = await unitOfWork.modelRepository.findById(modelId);
		if (!model) {
			throw new Error(`Model with id ${modelId} does not exist`);
		}
		if (!(await this.modelAuthorizationRuleset.canWrite(context, model))) {
			throw new Error("You do not have permission to delete this model");
		}
		const games = await unitOfWork.gameRepository.find([new FilterCondition("models", {_id: modelId})]);
		for (let game of games) {
			const positionedModel = game.models.find((otherModel) => otherModel.model === model._id);
			game.models = game.models.filter((positionedModel) => positionedModel.model !== model._id);
			await unitOfWork.gameRepository.update(game);
			await this.eventPublisher.publish(GAME_MODEL_DELETED, {
				gameId: game._id.toString(),
				gameModelDeleted: positionedModel,
			});
		}
		await unitOfWork.modelRepository.delete(model);
		if (await this.cache.exists(model.fileName)) {
			await this.cache.delete(model.fileName);
		}
		await this.authorizationService.cleanUpPermissions(modelId, unitOfWork);
		await unitOfWork.commit();
		return model;
	};

	getModels = async (context: SecurityContext, worldId: string): Promise<Model[]> => {
		const models = await this.modelRepository.find([new FilterCondition("world", worldId)]);
		const returnModels = [];
		for (let model of models) {
			if (await this.modelAuthorizationRuleset.canRead(context, model)) {
				returnModels.push(model);
			}
		}
		return returnModels;
	};

	private filenameExists = async (filename: string) => {
		const unitOfWork = this.dbUnitOfWorkFactory();
		const models = await unitOfWork.modelRepository.find([
			new FilterCondition("filename", filename),
		]);
		return models.length > 1;
	};
}
