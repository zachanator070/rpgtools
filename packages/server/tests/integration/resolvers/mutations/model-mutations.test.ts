import {defaultTestingContextFactory} from "../../DefaultTestingContextFactory";
import {FileUpload, Upload} from "graphql-upload";
import fs from "fs";
import {ModelApplicationService} from "../../../../src/services/model-application-service";
import {container} from "../../../../src/di/inversify";
import {INJECTABLE_TYPES} from "../../../../src/di/injectable-types";
import {CREATE_MODEL, DELETE_MODEL, UPDATE_MODEL} from "@rpgtools/common/src/gql-mutations";

process.env.TEST_SUITE = "model-mutations-test";

describe("model mutations", () => {
    const modelApplicationService: ModelApplicationService = container.get<ModelApplicationService>(INJECTABLE_TYPES.ModelService);
    let {
        server,
        mockSessionContextFactory,
        otherUser,
        otherUserSecurityContext,
        world,
        testRole,
        currentUser,
        testerSecurityContext,
        newFolder,
        otherPage,
        ...testingContext
    } = defaultTestingContextFactory();

    describe("with world", () => {

        const filename = "tests/integration/resolvers/mutations/pikachu.glb";
        const testFile: FileUpload = {
            encoding: "binary",
            mimetype: "application/gltf",
            filename: filename,
            createReadStream: () => fs.createReadStream(filename),
        };

        const testUpload = new Upload();
        testUpload.file = testFile;
        testUpload.promise = new Promise<FileUpload>((resolve) => {
            resolve(testFile);
        });

        beforeEach(async () => {
            ({
                mockSessionContextFactory,
                otherUser,
                otherUserSecurityContext,
                world,
                testRole,
                currentUser,
                testerSecurityContext,
                newFolder,
                otherPage,
            } = await testingContext.reset());
            mockSessionContextFactory.resetCurrentUser();
        });

        test("create model no permission", async () => {
            const result = await server.executeGraphQLQuery({
                query: CREATE_MODEL,
                variables: {name: "pikachu", file: testUpload, worldId: world._id, depth: 2, width: 1, height: 1, notes: 'pika pika'},
            });
            expect(result).toMatchSnapshot({
                errors: expect.arrayContaining([expect.any(Object)]),
            });
        });

        test("update model no permission", async () => {
            const model = await modelApplicationService.createModel(testerSecurityContext, world._id, "pikachu", testFile, 2, 1, 1, "pika pika");
            const result = await server.executeGraphQLQuery({
                query: UPDATE_MODEL,
                variables: {modelId: model._id, name: "new wiki", file: testUpload, worldId: world._id, depth: 2, width: 1, height: 1, notes: 'pika pika'},
            });
            expect(result).toMatchSnapshot({
                errors: expect.arrayContaining([expect.any(Object)]),
            });
        });

        test("delete model no permission", async () => {
            const model = await modelApplicationService.createModel(testerSecurityContext, world._id, "pikachu", testFile, 2, 1, 1, "pika pika");
            const result = await server.executeGraphQLQuery({
                query: DELETE_MODEL,
                variables: {modelId: model._id},
            });
            expect(result).toMatchSnapshot({
                errors: expect.arrayContaining([expect.any(Object)]),
            });
        });

        describe("with authenticated user", () => {
            beforeEach(async () => {
                mockSessionContextFactory.setCurrentUser(currentUser);
            });

            test("create model", async () => {
                const result = await server.executeGraphQLQuery({
                    query: CREATE_MODEL,
                    variables: {name: "new wiki", file: testUpload, worldId: world._id, depth: 2, width: 1, height: 1, notes: 'pika pika'},
                });
                expect(result).toMatchSnapshot({
                    data: {
                        createModel: {
                            _id: expect.any(String)
                        }
                    },
                    errors: undefined
                });
            });

            test("create two models with same filename", async () => {
                const result = await server.executeGraphQLQuery({
                    query: CREATE_MODEL,
                    variables: {name: "new wiki", file: testUpload, worldId: world._id, depth: 2, width: 1, height: 1, notes: 'pika pika'},
                });
                expect(result).toMatchSnapshot({
                    data: {
                        createModel: {
                            _id: expect.any(String)
                        }
                    },
                    errors: undefined
                });
                const secondResult = await server.executeGraphQLQuery({
                    query: CREATE_MODEL,
                    variables: {name: "new wiki", file: testUpload, worldId: world._id, depth: 2, width: 1, height: 1, notes: 'pika pika'},
                });
                expect(secondResult).toMatchSnapshot({
                    data: {
                        createModel: {
                            _id: expect.any(String)
                        }
                    },
                    errors: undefined
                });
            });

            test("update model", async () => {
                const model = await modelApplicationService.createModel(testerSecurityContext, world._id, "pikachu", testFile, 2, 1, 1, "pika pika");
                const result = await server.executeGraphQLQuery({
                    query: UPDATE_MODEL,
                    variables: {modelId: model._id, name: "new wiki", file: testUpload, worldId: world._id, depth: 2, width: 1, height: 1, notes: 'pika pika'},
                });
                expect(result).toMatchSnapshot({
                    data: {
                        updateModel: {
                            _id: expect.any(String),
                            fileId: expect.any(String),
                            accessControlList: expect.arrayContaining([
                                expect.objectContaining({
                                    _id: expect.any(String),
                                    subject: {
                                        _id: expect.any(String),
                                    },
                                    users: expect.arrayContaining([
                                        expect.objectContaining({
                                            _id: expect.any(String),
                                        })
                                    ]),
                                })
                            ])
                        }
                    },
                    errors: undefined
                });
            });

            test("delete model", async () => {
                const model = await modelApplicationService.createModel(testerSecurityContext, world._id, "pikachu", testFile, 2, 1, 1, "pika pika");
                const result = await server.executeGraphQLQuery({
                    query: DELETE_MODEL,
                    variables: {modelId: model._id},
                });
                expect(result).toMatchSnapshot({
                    data: {
                        deleteModel: {
                            _id: expect.any(String)
                        }
                    },
                    errors: undefined
                });
            });
        });
    });
});
