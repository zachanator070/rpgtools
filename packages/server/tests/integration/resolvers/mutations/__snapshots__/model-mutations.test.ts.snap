// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model mutations with world create model no permission 1`] = `
Object {
  "data": null,
  "errors": ArrayContaining [
    Any<Object>,
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world delete model no permission 1`] = `
Object {
  "data": null,
  "errors": ArrayContaining [
    Any<Object>,
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world update model no permission 1`] = `
Object {
  "data": null,
  "errors": ArrayContaining [
    Any<Object>,
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world with authenticated user create model 1`] = `
Object {
  "data": Object {
    "createModel": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world with authenticated user create two models with same filename 1`] = `
Object {
  "data": Object {
    "createModel": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world with authenticated user create two models with same filename 2`] = `
Object {
  "data": Object {
    "createModel": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world with authenticated user delete model 1`] = `
Object {
  "data": Object {
    "deleteModel": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`model mutations with world with authenticated user update model 1`] = `
Object {
  "data": Object {
    "updateModel": Object {
      "_id": Any<String>,
      "accessControlList": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "subject": Object {
            "_id": Any<String>,
          },
          "users": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
        },
      ],
      "canAdmin": true,
      "canWrite": true,
      "depth": 2,
      "fileId": Any<String>,
      "fileName": "tests/integration/resolvers/mutations/pikachu.glb",
      "height": 1,
      "name": "new wiki",
      "notes": "pika pika",
      "width": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
